File: Microsoft.DotNet.Try.Protocol.ClientApi/Project.cs
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

using System;
using System.Collections.Generic;
using System.Linq;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace Microsoft.DotNet.Try.Protocol.ClientApi
{
    [JsonConverter(typeof(ProjectJsonConverter))]
    public class Project : FeatureContainer
    {
        public string ProjectTemplate { get; }

        public SourceFile[] Files { get; }


        public Project(string projectTemplate, IEnumerable<SourceFile> files)
        {
            if (string.IsNullOrWhiteSpace(projectTemplate))
            {
                throw new ArgumentException("Value cannot be null or whitespace.", nameof(projectTemplate));
            }

            if (files == null)
            {
                throw new ArgumentNullException(nameof(files));
            }

            Files = files.Where(f => f != null).ToArray();

            if (Files.Length == 0)
            {
                throw new ArgumentException("Collection cannot be empty.", nameof(files));
            }


            ProjectTemplate = projectTemplate;

        }

        private class ProjectJsonConverter : FeatureContainerConverter<Project>
        {
            protected override void AddProperties(Project container, JObject o)
            {
                o.Add(new JProperty("projectTemplate", container.ProjectTemplate));
                o.Add(new JProperty("files", JArray.FromObject(container.Files)));
            }
        }
    }
}

File: MLS.Trydotnet.js/src/project.ts
// Copyright (c) .NET Foundation and contributors. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

export type Project = {
    package: string,
    packageVersion?:string,
    files: SourceFile[],
    [key: string]: any
};

export type SourceFile = {
    name: string,
    content: string,
};

export type SourceFileRegion = {
    id: string,
    content: string,
};

export function createProject(packageName: string, files: SourceFile[], usings?: string[]): Promise<Project> {
    if (!packageName || packageName.length === 0) {
        throw new Error("packageName can not be null or empty");
    }

    if (!files || files.length === 0) {
        throw new Error("at least a file is required");
    }

    let project: Project = {
        package: packageName,
        files: JSON.parse(JSON.stringify(files))
    };

    if (usings) {
        project.usings = JSON.parse(JSON.stringify(usings));
    }

    return Promise.resolve(project);
}
